src/directory_and_file_reader.c

```diff
@@ -0,0 +1,289 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+#include <errno.h>
+
+#include "rcutils/directory_and_file_reader.h"
+#include "rcutils/error_handling.h"
+#include "rcutils/filesystem.h"
+#include "rcutils/strdup.h"
+
+rcutils_dir_t
+rcutils_get_zero_initialized_dir(void)
+{
+  rcutils_dir_t zero_initialized_dir;
+  zero_initialized_dir.path = NULL;
+  zero_initialized_dir.has_next = 0;
+  zero_initialized_dir.dir = NULL;
+  #ifndef _WIN32
+  zero_initialized_dir.f = NULL;
+  #endif
+  zero_initialized_dir.allocator = rcutils_get_zero_initialized_allocator();
+  return zero_initialized_dir;
+}
+
+rcutils_file_t
+rcutils_get_zero_initialized_file(void)
+{
+  rcutils_file_t zero_initialized_file;
+  zero_initialized_file.path = NULL;
+  zero_initialized_file.name = NULL;
+  zero_initialized_file.is_dir = 0;
+  return zero_initialized_file;
+}
+
+rcutils_ret_t
+rcutils_file_fini(rcutils_file_t * file, rcutils_allocator_t allocator)
+{
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(file, RCUTILS_RET_INVALID_ARGUMENT);
+
+  if (file->path != NULL) {
+    allocator.deallocate(file->path, allocator.state);
+  }
+  if (file->name != NULL) {
+    allocator.deallocate(file->name, allocator.state);
+  }
+
+  return RCUTILS_RET_OK;
+}
+
+rcutils_ret_t
+rcutils_open_dir(rcutils_dir_t * dir, const char * path, rcutils_allocator_t allocator)
+{
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(dir, RCUTILS_RET_INVALID_ARGUMENT);
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(path, RCUTILS_RET_INVALID_ARGUMENT);
+
+  if (strlen(path) == 0) {
+    RCUTILS_SET_ERROR_MSG("The size of the path is 0");
+    return RCUTILS_RET_INVALID_ARGUMENT;
+  }
+
+  if (strlen(path) >= RCUTILS_DIR_PATH_MAX) {
+    RCUTILS_SET_ERROR_MSG("The size of the path is too long");
+    return RCUTILS_RET_INVALID_ARGUMENT;
+  }
+
+  dir->allocator = allocator;
+
+  if (dir->path != NULL) {
+    dir->allocator.deallocate(dir->path, dir->allocator.state);
+  }
+
+  if (dir->dir != NULL) {
+    errno = 0;
+    #ifndef _WIN32
+    int error = closedir(dir->dir);
+    // The closedir() function returns 0 on success
+    if (error) {
+      #else
+    // If the function succeeds, the return value is nonzero.
+    if (!FindClose(dir->dir)) {
+    #endif
```

Comment:
GitHub is failing to render this, maybe that's the source of your uncrustify issue too?